module/zfs/zfs_sysfs.c

```diff
@@ -0,0 +1,441 @@
+};
+
+#define	ZFS_FEATURE_COUNT	(sizeof (zfs_features) / sizeof (char *))
+
+/*
+ * A zfs_mod_kobj_t represents a zfs kobject in '/sys/module/zfs'
+ */
+typedef struct zfs_mod_kobj {
+	struct kobject		zko_kobj;
+	struct kobj_type	zko_kobj_type;
+	struct sysfs_ops	zko_sysfs_ops;
+	size_t			zko_attr_count;
+	struct attribute	*zko_attr_list;		/* allocated */
+	struct attribute	**zko_default_attrs;	/* allocated */
+} zfs_mod_kobj_t;
+
+#define	ATTR_TABLE_SIZE(cnt)	(sizeof (struct attribute) * (cnt))
+/* Note +1 for NULL terminator slot */
+#define	DEFAULT_ATTR_SIZE(cnt)	(sizeof (struct attribute *) * (cnt + 1))
+
+static zfs_mod_kobj_t runtime_features_kobj;
+static zfs_mod_kobj_t pool_features_kobj;
+static zfs_mod_kobj_t dataset_props_kobj;
+static zfs_mod_kobj_t pool_props_kobj;
+
+/*
+ * The show function is used to provide the content
+ * of an attribute into a PAGE_SIZE buffer.
+ */
+typedef ssize_t	(*sysfs_show_func)(struct kobject *, struct attribute *,
+    char *);
+
+static void
+zfs_kobj_release(struct kobject *kobj)
+{
+	zfs_mod_kobj_t *zkobj = container_of(kobj, zfs_mod_kobj_t, zko_kobj);
+
+	ASSERT(zkobj->zko_attr_count != 0);
+	ASSERT(zkobj->zko_attr_list);
+	ASSERT(zkobj->zko_default_attrs);
+
+	kmem_free(zkobj->zko_attr_list, ATTR_TABLE_SIZE(zkobj->zko_attr_count));
+	kmem_free(zkobj->zko_default_attrs,
+	    DEFAULT_ATTR_SIZE(zkobj->zko_attr_count));
+	zkobj->zko_attr_list = NULL;
+	zkobj->zko_default_attrs = NULL;
+}
+
+static void
+zfs_kobj_add_attr(zfs_mod_kobj_t *zkobj, int attr_num, const char *attr_name)
+{
+	ASSERT(attr_num < zkobj->zko_attr_count);
+	ASSERT(zkobj->zko_attr_list);
+	ASSERT(zkobj->zko_default_attrs);
+
+	zkobj->zko_attr_list[attr_num].name = attr_name;
+	zkobj->zko_attr_list[attr_num].mode = 0444;
+	zkobj->zko_default_attrs[attr_num] = &zkobj->zko_attr_list[attr_num];
+}
+
+static int
+zfs_kobj_init(zfs_mod_kobj_t *zkobj, int attr_cnt, sysfs_show_func show_func)
+{
+	zkobj->zko_attr_list = kmem_zalloc(ATTR_TABLE_SIZE(attr_cnt), KM_SLEEP);
+	if (zkobj->zko_attr_list == NULL)
+		return (ENOMEM);
+
+	zkobj->zko_default_attrs = kmem_zalloc(DEFAULT_ATTR_SIZE(attr_cnt),
+	    KM_SLEEP);
+	if (zkobj->zko_default_attrs == NULL) {
+		kmem_free(zkobj->zko_attr_list, ATTR_TABLE_SIZE(attr_cnt));
+		return (ENOMEM);
+	}
+
+	zkobj->zko_attr_count = attr_cnt;
+	zkobj->zko_sysfs_ops.show = show_func;
+	zkobj->zko_kobj_type.sysfs_ops = &zkobj->zko_sysfs_ops;
+	zkobj->zko_kobj_type.release = zfs_kobj_release;
+	zkobj->zko_kobj_type.default_attrs = zkobj->zko_default_attrs;
+
+	return (0);
+}
+
+static int
+zfs_kobj_add(zfs_mod_kobj_t *zkobj, struct kobject *parent, const char *name)
+{
+	/* zko_default_attrs must be NULL terminated */
+	ASSERT(zkobj->zko_default_attrs[zkobj->zko_attr_count] == NULL);
+
+	kobject_init(&zkobj->zko_kobj, &zkobj->zko_kobj_type);
+
+	int err = kobject_add(&zkobj->zko_kobj, parent, name);
+	if (err) {
+		printk(KERN_WARNING "%s: kobject_add %s failed (%d)\n",
+		    __func__, name, err);
```

Comment:
I believe the warning here is redundant, `kobject_add()` appears to already log to the console on error.

Reply:
good point, I'll drop it then